In [2]:
import pandas as pd
import numpy as np

In [4]:
file=pd.read_csv('road_accident.csv')

In [6]:
file

,accident_index,accident_date,day_of_week,junction_control,junction_detail,accident_severity,light_conditions,local_authority,carriageway_hazards,number_of_casualties,number_of_vehicles,police_force,road_surface_conditions,road_type,speed_limit,time,urban_or_rural_area,weather_conditions,vehicle_type
0,BS0000001,01-01-2021,Thursday,Give way or uncontrolled,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,2,Metropolitan Police,Dry,One way street,30,15:11,Urban,Fine no high winds,Car
1,BS0000002,05-01-2021,Monday,Give way or uncontrolled,Crossroads,Serious,Daylight,Kensington and Chelsea,NaN,11,2,Metropolitan Police,Wet or damp,Single carriageway,30,10:59,Urban,Fine no high winds,Taxi/Private hire car
2,BS0000003,04-01-2021,Sunday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Kensington and Chelsea,NaN,1,2,Metropolitan Police,Dry,Single carriageway,30,14:19,Urban,Fine no high winds,Taxi/Private hire car
3,BS0000004,05-01-2021,Monday,Auto traffic signal,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,2,Metropolitan Police,Frost or ice,Single carriageway,30,08:10,Urban,Other,Motorcycle over 500cc
4,BS0000005,06-01-2021,Tuesday,Auto traffic signal,Crossroads,Serious,Darkness - lights lit,Kensington and Chelsea,NaN,1,2,Metropolitan Police,Dry,Single carriageway,30,17:25,Urban,Fine no high winds,Car
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
307968,BS0307969,18-02-2022,Thursday,Data missing or out of range,Not at junction or within 20 metres,Slight,Daylight,Highland,NaN,2,1,Northern,Dry,Single carriageway,60,07:00,Rural,Fine no high winds,Car
307969,BS0307970,21-02-2022,Sunday,Data missing or out of range,Not at junction or within 20 metres,Slight,Darkness - no lighting,Highland,NaN,1,1,Northern,Frost or ice,Single carriageway,60,03:00,Rural,Fine no high winds,Car
307970,BS0307971,23-02-2022,Tuesday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Highland,NaN,1,3,Northern,Frost or ice,Single carriageway,30,09:38,Rural,Fine no high winds,Car
307971,BS0307972,23-02-2022,Tuesday,Give way or uncontrolled,T or staggered junction,Serious,Darkness - no lighting,Highland,NaN,1,2,Northern,Wet or damp,Single carriageway,60,18:25,Rural,Fine no high winds,Motorcycle over 500cc


In [10]:
file['accident_date']=pd.to_datetime(file['accident_date'],format='%d-%m-%Y')

In [12]:
file.dtypes.to_frame()

,0
accident_index,object
accident_date,datetime64[ns]
day_of_week,object
junction_control,object
junction_detail,object
accident_severity,object
light_conditions,object
local_authority,object
carriageway_hazards,object
number_of_casualties,int64


In [14]:
file.shape

(307973, 19)

In [16]:
file.size

5851487

- contains 5,851,487 which are not assured that the datas would be valid or not

In [19]:
file.describe()

,accident_date,number_of_casualties,number_of_vehicles,speed_limit
count,307973,307973.000000,307973.000000,307973.000000
mean,2021-12-23 22:19:39.804722944,1.356882,1.829063,38.866037
min,2021-01-01 00:00:00,1.000000,1.000000,10.000000
25%,2021-06-28 00:00:00,1.000000,1.000000,30.000000
50%,2021-12-08 00:00:00,1.000000,2.000000,30.000000
75%,2022-06-25 00:00:00,1.000000,2.000000,50.000000
max,2022-12-31 00:00:00,48.000000,32.000000,70.000000
std,NaN,0.815857,0.710477,14.032933


- std dev for no of casualites is .81 - which is less count of values 
- std dev for no of vehicles .71 - lesser value over all
- std dev for speed_limit with higher standard deviation or 14 - this may be resultant of an accident

In [ ]:
file['time']=pd.to_datetime(file['time'],format='%H:%M').dt.time

In [50]:
file.dtypes.to_frame()

,0
accident_index,object
accident_date,datetime64[ns]
day_of_week,object
junction_control,object
junction_detail,object
accident_severity,object
light_conditions,object
local_authority,object
carriageway_hazards,object
number_of_casualties,int64


In [52]:
# seperate columns for year,monthname,dayname
file['year']=file['accident_date'].dt.year

In [54]:
file['month_name']=file['accident_date'].dt.month_name()
file['day_name']=file['accident_date'].dt.day_name()

In [56]:
file

,accident_index,accident_date,day_of_week,junction_control,junction_detail,accident_severity,light_conditions,local_authority,carriageway_hazards,number_of_casualties,...,road_surface_conditions,road_type,speed_limit,time,urban_or_rural_area,weather_conditions,vehicle_type,year,month_name,day_name
0,BS0000001,2021-01-01,Thursday,Give way or uncontrolled,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,...,Dry,One way street,30,15:11:00,Urban,Fine no high winds,Car,2021,January,Friday
1,BS0000002,2021-01-05,Monday,Give way or uncontrolled,Crossroads,Serious,Daylight,Kensington and Chelsea,NaN,11,...,Wet or damp,Single carriageway,30,10:59:00,Urban,Fine no high winds,Taxi/Private hire car,2021,January,Tuesday
2,BS0000003,2021-01-04,Sunday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Kensington and Chelsea,NaN,1,...,Dry,Single carriageway,30,14:19:00,Urban,Fine no high winds,Taxi/Private hire car,2021,January,Monday
3,BS0000004,2021-01-05,Monday,Auto traffic signal,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,...,Frost or ice,Single carriageway,30,08:10:00,Urban,Other,Motorcycle over 500cc,2021,January,Tuesday
4,BS0000005,2021-01-06,Tuesday,Auto traffic signal,Crossroads,Serious,Darkness - lights lit,Kensington and Chelsea,NaN,1,...,Dry,Single carriageway,30,17:25:00,Urban,Fine no high winds,Car,2021,January,Wednesday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
307968,BS0307969,2022-02-18,Thursday,Data missing or out of range,Not at junction or within 20 metres,Slight,Daylight,Highland,NaN,2,...,Dry,Single carriageway,60,07:00:00,Rural,Fine no high winds,Car,2022,February,Friday
307969,BS0307970,2022-02-21,Sunday,Data missing or out of range,Not at junction or within 20 metres,Slight,Darkness - no lighting,Highland,NaN,1,...,Frost or ice,Single carriageway,60,03:00:00,Rural,Fine no high winds,Car,2022,February,Monday
307970,BS0307971,2022-02-23,Tuesday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Highland,NaN,1,...,Frost or ice,Single carriageway,30,09:38:00,Rural,Fine no high winds,Car,2022,February,Wednesday
307971,BS0307972,2022-02-23,Tuesday,Give way or uncontrolled,T or staggered junction,Serious,Darkness - no lighting,Highland,NaN,1,...,Wet or damp,Single carriageway,60,18:25:00,Rural,Fine no high winds,Motorcycle over 500cc,2022,February,Wednesday


In [62]:
# CY_Casualties (Current Year Casualties)
file['year'].value_counts()

year
2021    163554
2022    144419
Name: count, dtype: int64

In [66]:
current_year=file['year'].max()
current_year

2022

In [70]:
current_year_casualities=file[file['year']==current_year]
current_year_casualities.shape

(144419, 22)

In [ ]:
- current year casualities of 2022 has the count of 1,44,419 accidents
- which us reduced as compared to the previous year

In [82]:
# CY – Fatal Casualties - 2022
file['accident_severity'].value_counts().to_frame()

,count
accident_severity,
Slight,263280
Serious,40740
Fatal,3953


In [96]:
fatal=file[(file['year']==2022) & (file['accident_severity']=='Fatal')]['number_of_casualties'].sum()
print(f"Current year fatal casualities are {fatal} nos")

Current year fatal casualities are 2855 nos


In [104]:
# CY – Serious Casualties -2022
serious=file[(file['year']==2022)&(file['accident_severity']=='Serious')]['number_of_casualties'].sum()
print(f"Current year serious casualities are {serious} nos")

Current year serious casualities are 27045 nos


In [106]:
# CY – Slight Casualties – 2022
slight=file[(file['year']==2022)&(file['accident_severity']=='Slight')]['number_of_casualties'].sum()
print(f"Current year slight casualties are {slight} nos")

Current year slight casualties are 165837 nos


In [112]:
# Total Number of [Slight, Fatal, Serious] Casualties
total_casualties=(
    file.groupby('accident_severity')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)
total_casualties

,accident_severity,number_of_casualties
0,Slight,351436
1,Serious,59312
2,Fatal,7135


In [114]:
# Percentage(%) of Accidents that got Severity – Slight
slight_percentage=(
    file[file['accident_severity'].str.lower()=='slight']['number_of_casualties']
    .sum()
    /file['number_of_casualties'].sum()
)*100
print(f"Percentage of Slight accident severity is {round(slight_percentage,2)}")

Percentage of Slight accident severity is 84.1


In [116]:
# Percentage(%) of Accidents that got Severity – Fatal
fatal_percentage=(
    file[file['accident_severity'].str.lower()=='fatal']['number_of_casualties']
    .sum()
    /file['number_of_casualties'].sum()
)*100
print(f"Percentage of Slight accident severity is {round(fatal_percentage,2)}")

Percentage of Slight accident severity is 1.71


In [118]:
# Percentage(%) of Accidents that got Severity – Serious
serious_percentage=(
    file[file['accident_severity'].str.lower()=='serious']['number_of_casualties']
    .sum()
    /file['number_of_casualties'].sum()
)*100
print(f"Percentage of Slight accident severity is {round(serious_percentage,2)}")

Percentage of Slight accident severity is 14.19


In [122]:
# Vehicle Group – Total Number of Casualties
accident_by_vehicle=(
    file.groupby('vehicle_type')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .to_frame()
)
accident_by_vehicle

,number_of_casualties
vehicle_type,
Car,325922
Van / Goods 3.5 tonnes mgw or under,21298
Motorcycle over 500cc,15151
Bus or coach (17 or more pass seats),11710
Motorcycle 125cc and under,9109
Goods 7.5 tonnes mgw and over,8770
Taxi/Private hire car,7563
Motorcycle 50cc and under,4945
Motorcycle over 125cc and up to 500cc,4467


In [148]:
# method 2
def map_vechicles(v):
    if any(x in v for x in ['Car','Taxi/Private hire car']):
        return 'Cars'
    elif any(x in v for x in ['Motorcycle over 500cc','Motorcycle 125cc and under','Motorcycle 50cc and under','Motorcycle over 125cc and up to 500cc']):
        return 'Bike'
    elif any(x in v for x in ['Van / Goods 3.5 tonnes mgw or under','Bus or coach (17 or more pass seats)','Goods 7.5 tonnes mgw and over','Goods 7.5 tonnes mgw and over','Goods over 3.5t. and under 7.5t']):
        return 'Van'
    elif any(x in v for x in ['Bus or coach (17 or more pass seats)','Minibus (8 - 16 passenger seats)']):
        return 'Bus'
    elif any(x in v for x in ['Agricultural vehicle']):
        return 'Agricultural vehicle'
    else:
        return 'Other vechicle'

In [150]:
file['vehicle_category']=file['vehicle_type'].map(map_vechicles)
file

,accident_index,accident_date,day_of_week,junction_control,junction_detail,accident_severity,light_conditions,local_authority,carriageway_hazards,number_of_casualties,...,road_type,speed_limit,time,urban_or_rural_area,weather_conditions,vehicle_type,year,month_name,day_name,vehicle_category
0,BS0000001,2021-01-01,Thursday,Give way or uncontrolled,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,...,One way street,30,15:11:00,Urban,Fine no high winds,Car,2021,January,Friday,Cars
1,BS0000002,2021-01-05,Monday,Give way or uncontrolled,Crossroads,Serious,Daylight,Kensington and Chelsea,NaN,11,...,Single carriageway,30,10:59:00,Urban,Fine no high winds,Taxi/Private hire car,2021,January,Tuesday,Cars
2,BS0000003,2021-01-04,Sunday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Kensington and Chelsea,NaN,1,...,Single carriageway,30,14:19:00,Urban,Fine no high winds,Taxi/Private hire car,2021,January,Monday,Cars
3,BS0000004,2021-01-05,Monday,Auto traffic signal,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,...,Single carriageway,30,08:10:00,Urban,Other,Motorcycle over 500cc,2021,January,Tuesday,Bike
4,BS0000005,2021-01-06,Tuesday,Auto traffic signal,Crossroads,Serious,Darkness - lights lit,Kensington and Chelsea,NaN,1,...,Single carriageway,30,17:25:00,Urban,Fine no high winds,Car,2021,January,Wednesday,Cars
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
307968,BS0307969,2022-02-18,Thursday,Data missing or out of range,Not at junction or within 20 metres,Slight,Daylight,Highland,NaN,2,...,Single carriageway,60,07:00:00,Rural,Fine no high winds,Car,2022,February,Friday,Cars
307969,BS0307970,2022-02-21,Sunday,Data missing or out of range,Not at junction or within 20 metres,Slight,Darkness - no lighting,Highland,NaN,1,...,Single carriageway,60,03:00:00,Rural,Fine no high winds,Car,2022,February,Monday,Cars
307970,BS0307971,2022-02-23,Tuesday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Highland,NaN,1,...,Single carriageway,30,09:38:00,Rural,Fine no high winds,Car,2022,February,Wednesday,Cars
307971,BS0307972,2022-02-23,Tuesday,Give way or uncontrolled,T or staggered junction,Serious,Darkness - no lighting,Highland,NaN,1,...,Single carriageway,60,18:25:00,Rural,Fine no high winds,Motorcycle over 500cc,2022,February,Wednesday,Bike


In [152]:
acc_by_vehicle=(
    file.groupby('vehicle_category')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .to_frame()
)
acc_by_vehicle

,number_of_casualties
vehicle_category,
Cars,333485
Van,45182
Bike,33672
Other vechicle,3424
Bus,1088
Agricultural vehicle,1032


- Car has the high casuality count of 325922

In [130]:
# CY – Casualties Monthly Trend
casuality_by_month=(
    file.groupby(['month_name','year'])['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .to_frame()
)
casuality_by_month

,,number_of_casualties
month_name,year,
November,2021,20975
October,2021,20109
July,2021,19682
May,2021,18852
August,2021,18797
June,2021,18728
December,2021,18576
September,2021,18456
November,2022,18439


- november month in the year 2021 has the highest accident count of 20975

In [126]:
file['number_of_casualties'].sum()

417883

In [168]:
# Types of Road – Total Number of Casualties:
accident_by_roadtype=(
    file.groupby('road_type')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .to_frame()
)
accident_by_roadtype

,number_of_casualties
road_type,
Single carriageway,309698
Dual carriageway,67368
Roundabout,26828
One way street,7389
Slip road,6600


In [243]:
# Area – wise Percentage(%) and Total Number of Casualties
area_wise_count=(
    file.groupby('urban_or_rural_area')['number_of_casualties'].sum()
    .sort_values(ascending=False)
    .to_frame(name='count')
)
area_wise_percent=(
    file.groupby('urban_or_rural_area')['number_of_casualties'].sum()/file['number_of_casualties'].sum()*100
).to_frame(name='percent')

In [245]:
df1=pd.concat([area_wise_count,area_wise_percent],axis=1)
df1

,count,percent
urban_or_rural_area,,
Urban,255864,61.228621
Rural,162019,38.771379


In [212]:
def map_light(v):
    if any(x in v for x in ['Darkness - lights lit','Darkness - no lighting','Darkness - lighting unknown','Darkness - lights unlit']):
        return 'Night'
    else :
        return 'Day'

In [214]:
file['light_type']=file['light_conditions'].map(map_light)
file

,accident_index,accident_date,day_of_week,junction_control,junction_detail,accident_severity,light_conditions,local_authority,carriageway_hazards,number_of_casualties,...,speed_limit,time,urban_or_rural_area,weather_conditions,vehicle_type,year,month_name,day_name,vehicle_category,light_type
0,BS0000001,2021-01-01,Thursday,Give way or uncontrolled,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,...,30,15:11:00,Urban,Fine no high winds,Car,2021,January,Friday,Cars,Day
1,BS0000002,2021-01-05,Monday,Give way or uncontrolled,Crossroads,Serious,Daylight,Kensington and Chelsea,NaN,11,...,30,10:59:00,Urban,Fine no high winds,Taxi/Private hire car,2021,January,Tuesday,Cars,Day
2,BS0000003,2021-01-04,Sunday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Kensington and Chelsea,NaN,1,...,30,14:19:00,Urban,Fine no high winds,Taxi/Private hire car,2021,January,Monday,Cars,Day
3,BS0000004,2021-01-05,Monday,Auto traffic signal,T or staggered junction,Serious,Daylight,Kensington and Chelsea,NaN,1,...,30,08:10:00,Urban,Other,Motorcycle over 500cc,2021,January,Tuesday,Bike,Day
4,BS0000005,2021-01-06,Tuesday,Auto traffic signal,Crossroads,Serious,Darkness - lights lit,Kensington and Chelsea,NaN,1,...,30,17:25:00,Urban,Fine no high winds,Car,2021,January,Wednesday,Cars,Night
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
307968,BS0307969,2022-02-18,Thursday,Data missing or out of range,Not at junction or within 20 metres,Slight,Daylight,Highland,NaN,2,...,60,07:00:00,Rural,Fine no high winds,Car,2022,February,Friday,Cars,Day
307969,BS0307970,2022-02-21,Sunday,Data missing or out of range,Not at junction or within 20 metres,Slight,Darkness - no lighting,Highland,NaN,1,...,60,03:00:00,Rural,Fine no high winds,Car,2022,February,Monday,Cars,Night
307970,BS0307971,2022-02-23,Tuesday,Give way or uncontrolled,T or staggered junction,Slight,Daylight,Highland,NaN,1,...,30,09:38:00,Rural,Fine no high winds,Car,2022,February,Wednesday,Cars,Day
307971,BS0307972,2022-02-23,Tuesday,Give way or uncontrolled,T or staggered junction,Serious,Darkness - no lighting,Highland,NaN,1,...,60,18:25:00,Rural,Fine no high winds,Motorcycle over 500cc,2022,February,Wednesday,Bike,Night


In [236]:
# Count of Casualties By Light Conditions
casualties_by_light=(
    file.groupby('light_type')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .to_frame(name='count')
)
casualties_by_light

,count
light_type,
Day,304963
Night,112920


In [240]:
# 14. Percentage (%) and Segregation of Casualties by Different Light Conditions
casualties_percent=(
    file.groupby('light_type')['number_of_casualties']
    .sum()/file['number_of_casualties'].sum()*100
).to_frame(name='percentage')
df2=pd.concat([casualties_by_light,casualties_percent],axis=1)
df2

,count,percentage
light_type,,
Day,304963,72.978082
Night,112920,27.021918


In [210]:
# Top 10 Local Authority with Highest Total Number of Casualties
casuality_by_authority=(
    file.groupby('local_authority')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .to_frame()
)
casuality_by_authority.iloc[0:10]

,number_of_casualties
local_authority,
Birmingham,8611
Leeds,5821
Bradford,4431
Manchester,4366
Liverpool,4052
Cornwall,3820
Sheffield,3737
Kirklees,3312
County Durham,3295


In [228]:
casuality_by_authority=(
    file.groupby('local_authority')['number_of_casualties']
    .sum()
    .sort_values(ascending=False)
    .nlargest(10)
    .to_frame()
)
casuality_by_authority

,number_of_casualties
local_authority,
Birmingham,8611
Leeds,5821
Bradford,4431
Manchester,4366
Liverpool,4052
Cornwall,3820
Sheffield,3737
Kirklees,3312
County Durham,3295
